File: frontend/app/types/reservation.ts
export type ReservationStatus =
  | 'HELD'
  | 'CONFIRMED'
  | 'SEATED'
  | 'COMPLETED'
  | 'CANCELLED'
  | 'EXPIRED'

export type Reservation = {
  id: number
  restaurant_id: number
  table_id: number
  table_number?: string | null
  table_capacity?: number | null
  user_id?: number | null
  order_id?: number | null
  party_size: number
  starts_at: string
  ends_at: string
  status: ReservationStatus
  hold_expires_at?: string | null
  guest_name: string
  guest_email?: string | null
  guest_phone?: string | null
  notes?: string | null
  created_at: string
  /** Minutes a seated party has stayed past their booked end (0 = on time). */
  overdue_minutes?: number
  /** Name of the seated party still at this booking's table when it is due. */
  blocked_by?: string | null
}

export type AvailableTable = {
  id: number
  table_number: string
  capacity: number
  zone?: string | null
  status: string
}

/** A table on the customer seating plan, with its state for the searched time. */
export type FloorTable = {
  id: number
  table_number: string
  capacity: number
  zone?: string | null
  shape?: 'ROUND' | 'SQUARE' | 'RECT'
  pos_x?: number | null
  pos_y?: number | null
  /** AVAILABLE, UNAVAILABLE (taken at that time) or TOO_SMALL (for the party). */
  state: 'AVAILABLE' | 'UNAVAILABLE' | 'TOO_SMALL'
}

export type AvailabilityResponse = {
  starts_at: string
  ends_at: string
  party_size: number
  tables: AvailableTable[]
  /** Every active table with its state, for drawing the seating plan. */
  floor: FloorTable[]
  /** Nearby start times with a free table; only present when `tables` is empty. */
  suggested_times: string[]
}

export type ReservationConflict = {
  id: number
  guest_name: string
  party_size: number
  starts_at: string
  ends_at: string
  status: ReservationStatus
}

/** Slot status for a requested time — not the table's status right now. */
export type SlotStatus = 'AVAILABLE' | 'RESERVED' | 'OCCUPIED' | 'CLEANING' | 'TOO_SMALL'

export type AdminTableAvailability = {
  id: number
  table_number: string
  capacity: number
  zone?: string | null
  shape?: 'ROUND' | 'SQUARE' | 'RECT'
  pos_x?: number | null
  pos_y?: number | null
  floor_status: string
  slot_status: SlotStatus
  available: boolean
  conflicts: ReservationConflict[]
}

export type AdminAvailabilityResponse = {
  starts_at: string
  ends_at: string
  party_size: number
  tables: AdminTableAvailability[]
}

export type TableReservationBrief = {
  id: number
  guest_name: string
  guest_phone?: string | null
  party_size: number
  starts_at: string
  ends_at: string
  status: ReservationStatus
  /** True when this booking is why the table is RESERVED/OCCUPIED right now. */
  blocking: boolean
  overdue_minutes?: number
}

export type UpdateReservationPayload = Partial<{
  table_id: number
  party_size: number
  starts_at: string
  duration_minutes: number
  guest_name: string
  guest_email: string
  guest_phone: string
  notes: string
}>

export type CreateReservationPayload = {
  table_id: number
  party_size: number
  starts_at: string
  duration_minutes?: number
  guest_name: string
  guest_email?: string
  guest_phone?: string
  notes?: string
  hold?: boolean
}

File: backend/app/schemas/reservation.py
"""Reservation Pydantic schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.enums import ReservationStatus


class ReservationCreate(BaseModel):
    table_id: int
    party_size: int = Field(ge=1, le=20)
    starts_at: datetime
    duration_minutes: int = Field(default=90, ge=30, le=240)
    guest_name: str = Field(min_length=1, max_length=200)
    guest_email: EmailStr | None = None
    guest_phone: str | None = Field(default=None, max_length=30)
    notes: str | None = Field(default=None, max_length=1000)
    hold: bool = False  # True = soft hold (expires); False = confirm immediately


class AdminReservationCreate(ReservationCreate):
    user_id: int | None = None
    seat_immediately: bool = False


class ReservationStatusUpdate(BaseModel):
    status: ReservationStatus
    notes: str | None = Field(default=None, max_length=1000)


class ReservationExtend(BaseModel):
    """Give seated guests more time (they're running over)."""

    minutes: int = Field(ge=5, le=120)


class ReservationUpdate(BaseModel):
    """Admin edit / reschedule of an upcoming reservation. Omitted fields are unchanged."""

    table_id: int | None = None
    party_size: int | None = Field(default=None, ge=1, le=20)
    starts_at: datetime | None = None
    duration_minutes: int | None = Field(default=None, ge=30, le=240)
    guest_name: str | None = Field(default=None, min_length=1, max_length=200)
    guest_email: EmailStr | None = None
    guest_phone: str | None = Field(default=None, max_length=30)
    notes: str | None = Field(default=None, max_length=1000)


class ReservationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    restaurant_id: int
    table_id: int
    table_number: str | None = None
    table_capacity: int | None = None
    user_id: int | None
    order_id: int | None
    party_size: int
    starts_at: datetime
    ends_at: datetime
    status: ReservationStatus
    hold_expires_at: datetime | None
    guest_name: str
    guest_email: str | None
    guest_phone: str | None
    notes: str | None
    created_at: datetime
    # Seated guests still at the table past their booked end time (0 = not overdue).
    overdue_minutes: int = 0
    # For an upcoming booking: name of the seated party still occupying its table when it is due.
    blocked_by: str | None = None
    # The guest tapped "I'll be there" from the reminder email.
    guest_confirmed_at: datetime | None = None


class AvailableTableRead(BaseModel):
    id: int
    table_number: str
    capacity: int
    zone: str | None = None
    status: str


class ReservationConflict(BaseModel):
    """An existing booking that overlaps the requested slot."""

    id: int
    guest_name: str
    party_size: int
    starts_at: datetime
    ends_at: datetime
    status: ReservationStatus


class TableReservationBrief(BaseModel):
    id: int
    guest_name: str
    guest_phone: str | None = None
    party_size: int
    starts_at: datetime
    ends_at: datetime
    status: ReservationStatus
    # True when this booking is what makes the table RESERVED/OCCUPIED right now.
    blocking: bool = False
    overdue_minutes: int = 0


class AdminTableAvailability(BaseModel):
    id: int
    table_number: str
    capacity: int
    zone: str | None = None
    shape: str = "SQUARE"
    pos_x: float | None = None
    pos_y: float | None = None
    floor_status: str
    # AVAILABLE | RESERVED | OCCUPIED | CLEANING | TOO_SMALL — for the requested slot, not "now".
    slot_status: str
    available: bool
    conflicts: list[ReservationConflict] = Field(default_factory=list)


class FloorTableRead(BaseModel):
    """A table on the customer-facing floor plan, with its state for the searched slot."""

    id: int
    table_number: str
    capacity: int
    zone: str | None = None
    shape: str = "SQUARE"
    pos_x: float | None = None
    pos_y: float | None = None
    # AVAILABLE | UNAVAILABLE (taken at that time) | TOO_SMALL (for the party)
    state: str


class AvailabilityResponse(BaseModel):
    starts_at: datetime
    ends_at: datetime
    party_size: int
    tables: list[AvailableTableRead]
    # Every active table with its state, so the client can draw the seating plan.
    floor: list[FloorTableRead] = Field(default_factory=list)
    # Nearby start times that do have a free table (only filled when `tables` is empty).
    suggested_times: list[datetime] = Field(default_factory=list)


class AdminAvailabilityResponse(BaseModel):
    starts_at: datetime
    ends_at: datetime
    party_size: int
    tables: list[AdminTableAvailability]
